### PR TITLE
Vesting Escrow Wallet wrapper

### PR DESCRIPTION
**Pull request by shuffledex**

### CODE REVIEW

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 1]
```diff
@@ -0,0 +1,931 @@
...
+  valueToWei,
+  weiToValue,
+  dateToBigNumber,
+  bytes32ToString,
+  bytes32ArrayToStringArray,
+  bigNumberToDate,
+  stringToBytes32,
+} from '../../../utils/convert';
+
+const TRANSFER_SUCCESS = '0x51';
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: I'm pretty sure these codes are defined in an enum somewhere

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 2]
```diff
@@ -0,0 +1,931 @@
...
+  public removeTemplate = async (params: RemoveTemplateParams) => {
+    assert.assert(await this.isCallerAllowed(params.txData, Perm.Admin), 'Caller is not allowed');
+    assert.assert(params.name !== '', 'Invalid name');
+    assert.assert((await this.getAllTemplateNames()).includes(params.name), 'Template not found');
+    // TODO 3.1: require(templateToUsers[_name].length == 0, "Template is used");
+    return (await this.contract).removeTemplate.sendTransactionAsync(params.name, params.txData, params.safetyFactor);
+  };
+
+  /**
+   * Returns count of the templates those can be used for creating schedule
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: Change it to

`Returns the amount of templates that can be used for creating schedules`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 3]
```diff
@@ -0,0 +1,931 @@
...
+   * @return Count of the templates
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `@return Amount of templates`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 4]
```diff
@@ -0,0 +1,931 @@
...
+   */
+  public getTemplateCount = async () => {
+    const result = await (await this.contract).getTemplateCount.callAsync();
+    return result.toNumber();
+  };
+
+  /**
+   * Gets the list of the template names those can be used for creating schedule
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `Gets the list of template names that can be used for creating schedules`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 5]
```diff
@@ -0,0 +1,931 @@
...
+   * @return Array of all template names were created
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `@return Array of all template names`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 6]
```diff
@@ -0,0 +1,931 @@
...
+   */
+  public getAllTemplateNames = async () => {
+    const results = await (await this.contract).getAllTemplateNames.callAsync();
+    return bytes32ArrayToStringArray(results);
+  };
+
+  /**
+   * Adds vesting schedules for each of the beneficiary's address
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `Adds a vesting schedule for the beneficiary address`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 7]
```diff
@@ -0,0 +1,931 @@
...
+      numberToBigNumber(params.duration),
+      numberToBigNumber(params.frequency),
+      dateToBigNumber(params.startTime),
+      params.txData,
+      params.safetyFactor,
+    );
+  };
+
+  /**
+   * Adds vesting schedules from template for the beneficiary
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `Adds a vesting schedule for the beneficiary address from a template`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 8]
```diff
@@ -0,0 +1,931 @@
...
+      params.beneficiary,
+      stringToBytes32(params.templateName),
+      dateToBigNumber(params.startTime),
+      params.txData,
+      params.safetyFactor,
+    );
+  };
+
+  /**
+   * Modifies vesting schedules for each of the beneficiary
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `Modifies a vesting schedule for a beneficiary address`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 9]
```diff
@@ -0,0 +1,931 @@
...
+      params.beneficiary,
+      stringToBytes32(params.templateName),
+      dateToBigNumber(params.startTime),
+      params.txData,
+      params.safetyFactor,
+    );
+  };
+
+  /**
+   * Revokes vesting schedule with given template name for given beneficiary
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `Revokes a vesting schedule with a given template name for a given beneficiary`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 10]
```diff
@@ -0,0 +1,931 @@
...
+    return (await this.contract).revokeSchedule.sendTransactionAsync(
+      params.beneficiary,
+      stringToBytes32(params.templateName),
+      params.txData,
+      params.safetyFactor,
+    );
+  };
+
+  /**
+   * Revokes all vesting schedules for given beneficiary's address
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `Revokes all vesting schedules for a given beneficiary address`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 11]
```diff
@@ -0,0 +1,931 @@
...
+    assert.isNonZeroETHAddressHex('beneficiary', params.beneficiary);
+    return (await this.contract).revokeAllSchedules.sendTransactionAsync(
+      params.beneficiary,
+      params.txData,
+      params.safetyFactor,
+    );
+  };
+
+  /**
+   * Returns beneficiary's schedule created using template name
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `Returns a schedule with a given template name for a given beneficiary address`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 12]
```diff
@@ -0,0 +1,931 @@
...
+   * @return beneficiary's schedule data (numberOfTokens, duration, frequency, startTime, claimedTokens, State)
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `State` should be `state`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 13]
```diff
@@ -0,0 +1,931 @@
...
+   */
+  public getSchedule = async (params: GetScheduleParams) => {
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: Return type should be specified as `Schedule` (using the interface above)

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 14]
```diff
@@ -0,0 +1,931 @@
...
+      duration: result[1].toNumber(),
+      frequency: result[2].toNumber(),
+      startTime: bigNumberToDate(result[3]),
+      claimedTokens: result[4].toNumber(),
+      State: StateStatus[result[5].toNumber()],
+    };
+  };
+
+  /**
+   * Returns list of the template names for given beneficiary's address
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `Returns a list of the template names for a given beneficiary address`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 15]
```diff
@@ -0,0 +1,931 @@
...
+   * @return List of the template names that were used for schedule creation
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `@return List of template names used for the beneficiary address' schedules`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 16]
```diff
@@ -0,0 +1,931 @@
...
+   */
+  public getTemplateNames = async (params: GetTemplateNamesParams) => {
+    assert.isNonZeroETHAddressHex('beneficiary', params.beneficiary);
+    const result = await (await this.contract).getTemplateNames.callAsync(params.beneficiary);
+    return bytes32ArrayToStringArray(result);
+  };
+
+  /**
+   * Returns count of the schedules were created for given beneficiary
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `Returns amount of schedules for a given beneficiary address`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 17]
```diff
@@ -0,0 +1,931 @@
...
+   * @return Count of beneficiary's schedules
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `@return Amount of schedules`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 18]
```diff
@@ -0,0 +1,931 @@
...
+    return (await this.contract).pushAvailableTokensMulti.sendTransactionAsync(
+      numberToBigNumber(params.fromIndex),
+      numberToBigNumber(params.toIndex),
+      params.txData,
+      params.safetyFactor,
+    );
+  };
+
+  /**
+   * Used to bulk add vesting schedules for each of beneficiary
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `Used to bulk add vesting schedules for each of the beneficiaries`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 19]
```diff
@@ -0,0 +1,931 @@
...
+   * @return Count of beneficiary's schedules
+   */
+  public getScheduleCount = async (params: GetScheduleCountParams) => {
+    assert.isNonZeroETHAddressHex('beneficiary', params.beneficiary);
+    const result = await (await this.contract).getScheduleCount.callAsync(params.beneficiary);
+    return weiToValue(result, FULL_DECIMALS).toNumber();
+  };
+
+  /**
+   * Used to bulk send available tokens for each of the beneficiaries
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `Used to bulk send available tokens to each of the beneficiaries`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 20]
```diff
@@ -0,0 +1,931 @@
...
+      params.startTimes.map(startTime => {
+        return dateToBigNumber(startTime);
+      }),
+      params.txData,
+      params.safetyFactor,
+    );
+  };
+
+  /**
+   * Used to bulk add vesting schedules from template for each of the beneficiary
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `Used to bulk add vesting schedules from templates for each of the beneficiary addresses`

File: examples/vestingEscrowWallet.ts [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+    treasuryWallet: myAddress,
+    protocolVersion: '0',
+  });
+
+  // Create a Security Token Instance
+  const tickerSecurityTokenInstance = await polymathAPI.tokenFactory.getSecurityTokenInstanceFromTicker(ticker!);
+  console.log('ST address:', await tickerSecurityTokenInstance.address());
+
+  // Get permission manager factory address
+  const moduleStringName = 'VestingEscrowWallet';
```
[the remark this answers is not in the document]
Comment (shuffledex) [reply examples/vestingEscrowWallet.ts]: could I replace... `resultNames.indexOf(moduleStringName)` to `resultNames.indexOf(moduleName)` ?

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 22]
```diff
@@ -0,0 +1,931 @@
...
+/**
+ * @param name Name of the template will be created
+ * @param numberOfTokens Number of tokens that should be assigned to schedule
+ * @param duration Duration of the vesting schedule
+ * @param frequency Frequency of the vesting schedule
+ */
+interface AddTemplateParams extends TxParams {
+  name: string;
+  numberOfTokens: number;
+  duration: number;
```
[the remark this answers is not in the document]
Comment (shuffledex) [reply src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: I dont understand which param should be bignumber.

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 23]
```diff
@@ -0,0 +1,931 @@
...
+  frequency: number;
```
[the remark this answers is not in the document]
Comment (shuffledex) [reply src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: which one?

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 24]
```diff
@@ -21,6 +21,7 @@ import {
   BigNumber,
 } from '@polymathnetwork/abi-wrappers';
 import { schemas } from '@0x/json-schemas';
+import { StatusCodes } from '@0x/types';
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: Why is this import necessary?

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 25]
```diff
@@ -482,18 +490,18 @@ export default class VestingEscrowWalletWrapper extends ModuleWrapper {
...
     assert.assert(await this.isCallerAllowed(params.txData, Perm.Admin), 'Caller is not allowed');
 
-    assert.assert(params.numberOfTokens > 0, 'Number of tokens should be > 0');
+    assert.assert(params.numberOfTokens.toNumber() > 0, 'Number of tokens should be > 0');
 
     const canTransferFromResult = await (await this.securityTokenContract()).canTransferFrom.callAsync(
       await this.getCallerAddress(params.txData),
       await this.address(),
-      valueToWei(numberToBigNumber(params.numberOfTokens), FULL_DECIMALS),
+      params.numberOfTokens,
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: Shouldn't this still be converted to wei?

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 26]
```diff
@@ -482,18 +490,18 @@ export default class VestingEscrowWalletWrapper extends ModuleWrapper {
...
       '0x00',
     );
-    assert.assert(canTransferFromResult[0] === TRANSFER_SUCCESS, 'Failed transferFrom');
+    assert.assert(canTransferFromResult[0] === TransferStatusCode.TransferSuccess, 'Failed transferFrom');
 
     return (await this.contract).depositTokens.sendTransactionAsync(
-      valueToWei(numberToBigNumber(params.numberOfTokens), FULL_DECIMALS),
+      params.numberOfTokens,
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: Shouldn't this still be converted to wei?

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 27]
```diff
@@ -445,25 +450,28 @@ export default class VestingEscrowWalletWrapper extends ModuleWrapper {
...
-    return result.toNumber();
+    return result;
   };
 
   public schedules = async (params: SchedulesParams) => {
-    return (await this.contract).schedules.callAsync(params.beneficiary, numberToBigNumber(params.index));
+    const result = await (await this.contract).schedules.callAsync(params.beneficiary, numberToBigNumber(params.index));
+    const schedule: Schedule = {
+      templateName: bytes32ToString(result[0]),
+      claimedTokens: result[1].toNumber(),
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: Shouldn't this be converted from wei?

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 28]
```diff
@@ -445,25 +450,28 @@ export default class VestingEscrowWalletWrapper extends ModuleWrapper {
 
   public unassignedTokens = async () => {
     const result = await (await this.contract).unassignedTokens.callAsync();
-    return result.toNumber();
+    return result;
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: Shouldn't this be converted from wei?

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 29]
```diff
@@ -504,23 +512,19 @@ export default class VestingEscrowWalletWrapper extends ModuleWrapper {
...
+    assert.assert(params.amount.toNumber() > 0, 'Amount cannot be zero');
 
     const unassignedTokens = await this.unassignedTokens();
-    assert.assert(params.amount <= unassignedTokens, 'Amount is greater than unassigned tokens');
+    assert.assert(params.amount.isLessThanOrEqualTo(unassignedTokens), 'Amount is greater than unassigned tokens');
 
     const canTransferResult = await (await this.securityTokenContract()).canTransfer.callAsync(
       await this.getTreasuryWallet(),
-      numberToBigNumber(params.amount),
+      params.amount,
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: Shouldn't this be converted to wei?

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 30]
```diff
@@ -504,23 +512,19 @@ export default class VestingEscrowWalletWrapper extends ModuleWrapper {
...
     );
-    assert.assert(canTransferResult[0] === TRANSFER_SUCCESS, 'Transfer failed');
+    assert.assert(canTransferResult[0] === TransferStatusCode.TransferSuccess, 'Transfer failed');
 
-    return (await this.contract).sendToTreasury.sendTransactionAsync(
-      numberToBigNumber(params.amount),
-      params.txData,
-      params.safetyFactor,
-    );
+    return (await this.contract).sendToTreasury.sendTransactionAsync(params.amount, params.txData, params.safetyFactor);
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: Shouldn't the amount here be in wei?

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 31]
```diff
@@ -561,7 +565,7 @@ export default class VestingEscrowWalletWrapper extends ModuleWrapper {
     await this.validateTemplate(params.numberOfTokens, params.duration, params.frequency);
     return (await this.contract).addTemplate.sendTransactionAsync(
       stringToBytes32(params.name),
-      valueToWei(numberToBigNumber(params.numberOfTokens), FULL_DECIMALS),
+      params.numberOfTokens,
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: Shouldn't this still be converted to wei?

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 32]
```diff
@@ -610,7 +618,7 @@ export default class VestingEscrowWalletWrapper extends ModuleWrapper {
     return (await this.contract).addSchedule.sendTransactionAsync(
       params.beneficiary,
       stringToBytes32(params.templateName),
-      valueToWei(numberToBigNumber(params.numberOfTokens), FULL_DECIMALS),
+      params.numberOfTokens,
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: Shouldn't this still be converted to wei?

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 33]
```diff
@@ -680,26 +688,46 @@ export default class VestingEscrowWalletWrapper extends ModuleWrapper {
...
+        state = StateStatus.Started;
+        break;
+      default:
+        break;
+    }
 
-    return {
-      numberOfTokens: result[0].toNumber(),
+    const response: BeneficiarySchedule = {
+      numberOfTokens: result[0],
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: Shouldn't this be converted from wei?

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 34]
```diff
@@ -680,26 +688,46 @@ export default class VestingEscrowWalletWrapper extends ModuleWrapper {
...
     this.checkSchedule(params.beneficiary, params.templateName);
-    const result = await (await this.contract).getSchedule.callAsync(params.beneficiary, params.templateName);
+    const result = await (await this.contract).getSchedule.callAsync(
+      params.beneficiary,
+      stringToBytes32(params.templateName),
+    );
+
+    let state: StateStatus = StateStatus.Created;
+    switch (StateStatus[result[5].toNumber()]) {
+      case 'Created':
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: These cases should use the enum values

Example: `case StateStatus.Created`

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 35]
```diff
@@ -770,9 +798,7 @@ export default class VestingEscrowWalletWrapper extends ModuleWrapper {
       params.templateNames.map(name => {
         return stringToBytes32(name);
       }),
-      params.numberOfTokens.map(number => {
-        return valueToWei(numberToBigNumber(number), FULL_DECIMALS);
-      }),
+      params.numberOfTokens,
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: Shouldn't this still be converted to wei?

File: src/types.ts [hunk 36]
```diff
@@ -61,7 +61,9 @@ import {
   VolumeRestrictionTMEventArgs,
   VestingEscrowWalletContract,
   VestingEscrowWalletEvents,
-  VestingEscrowWalletUnpauseEventArgs,
```
Comment (monitz87) [inline src/types.ts]: Why aren't you exporting the event args?

Comment (shuffledex) [reply src/types.ts]: It was a typo mistake, I added "Unpause" but in types we should only import the generic EventArgs for each wrapper.

Comment (monitz87) [reply src/types.ts]: What I meant was, why isn't `VestingEscrowWalletEventArgs` being exported?

File: src/PolymathAPI.ts [hunk 37]
```diff
@@ -160,12 +159,10 @@ export class PolymathAPI {
...
-      ERC20Detailed.abi,
-      PolyToken.abi,
-      PolyTokenFaucet.abi,
-      ISecurityToken.abi,
-      // Wallet
-      VestingEscrowWallet.abi,
+      ERC20DetailedContract.ABI(),
+      PolyTokenContract.ABI(),
+      PolyTokenFaucetContract.ABI(),
+      ISecurityTokenContract.ABI(),
```
Comment (monitz87) [inline src/PolymathAPI.ts]: You should add the Vesting Escrow Wallet ABI to this array

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts (hunk 22 above)
[the remark this answers is not in the document]
Comment (monitz87) [reply src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: duration

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (monitz87) [reply src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: frequency

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 38]
```diff
@@ -0,0 +1,931 @@
...
+
+/**
+ * @param beneficiary Address of the beneficiary for whom it is scheduled
+ * @param templateName Name of the template that will be created
+ * @param numberOfTokens Total number of tokens for created schedule
+ * @param duration Duration of the created vesting schedule
+ * @param frequency Frequency of the created vesting schedule
+ * @param startTime Start time of the created vesting schedule
+ */
+interface AddScheduleParams extends TxParams {
```
[the remark this answers is not in the document]
Comment (monitz87) [reply src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `duration` and `frequency` should be BigNumber

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 39]
```diff
@@ -0,0 +1,931 @@
...
+
+/**
+ * @param beneficiaries Array of the beneficiary's addresses
+ * @param templateNames Array of the template names
+ * @param numberOfTokens Array of number of tokens should be assigned to schedules
+ * @param durations Array of the vesting duration
+ * @param frequencies Array of the vesting frequency
+ * @param startTimes Array of the vesting start time
+ */
+interface AddScheduleMultiParams extends TxParams {
```
[the remark this answers is not in the document]
Comment (monitz87) [reply src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `durations` and `frequencies` should be BigNumber[]

File: src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts [hunk 40]
```diff
@@ -0,0 +1,964 @@
...
+  Completed,
+}
+
+interface Schedule {
+  templateName: string;
+  claimedTokens: number;
+  startTime: Date;
+}
+
+interface BeneficiarySchedule {
```
Comment (monitz87) [inline src/contract_wrappers/modules/wallet/vesting_escrow_wallet_wrapper.ts]: `duration` and `frequency` should be BigNumber
